pkg/controller/handlers/mcpcatalog/mcpcatalog.go

```diff
@@ -273,3 +278,114 @@ func (h *Handler) SetUpDefaultMCPCatalog(ctx context.Context, c client.Client) e
+		if server.Spec.ToolReferenceName != "" || server.Spec.ThreadName != "" || server.Spec.SharedWithinMCPCatalogName != "" {
+			// For legacy gptscript tools, project-scoped servers, and multi-user servers created by the admin, we don't need to check them.
+			continue
+		}
+
+		user, err := h.gatewayClient.UserByID(req.Ctx, server.Spec.UserID)
+		if err != nil {
+			return fmt.Errorf("failed to get user %s: %w", server.Spec.UserID, err)
+		}
+
+		if user.Role.HasRole(types.RoleAdmin) {
+			// Don't delete servers created by admins.
+			continue
+		}
+
+		if server.Spec.MCPServerCatalogEntryName == "" {
+			// If the server doesn't have a catalog entry name, that's bad, because it should. Delete it.
+			log.Infof("Deleting MCP server %q because it does not correspond to a catalog entry", server.Name)
+			if err := req.Delete(&server); err != nil {
+				return fmt.Errorf("failed to delete MCP server %s: %w", server.Name, err)
+			}
+			continue
+		}
+
+		hasAccess, err := h.accessControlRuleHelper.UserHasAccessToMCPServerCatalogEntry(server.Spec.UserID, server.Spec.MCPServerCatalogEntryName)
+		if err != nil {
+			return fmt.Errorf("failed to check if user %s has access to catalog entry %s: %w", server.Spec.UserID, server.Spec.MCPServerCatalogEntryName, err)
+		}
+
+		if !hasAccess {
+			log.Infof("Deleting MCP server %q because it is no longer authorized to exist", server.Name)
+			if err := req.Delete(&server); err != nil {
+				return fmt.Errorf("failed to delete MCP server %s: %w", server.Name, err)
+			}
+		}
+	}
+
+	return nil
+}
+
+// DeleteUnauthorizedMCPServerInstances is a handler that deletes MCPServerInstances that point to multi-user MCPServers created by the admin,
+// where the user who owns the MCPServerInstance is no longer authorized to use the MCPServer.
+// This can happen whenever AccessControlRules change.
+func (h *Handler) DeleteUnauthorizedMCPServerInstances(req router.Request, _ router.Response) error {
+	// List AccessControlRules so that this handler gets triggered any time one of them changes.
+	if err := req.List(&v1.AccessControlRuleList{}, &client.ListOptions{
+		Namespace: system.DefaultNamespace,
+	}); err != nil {
+		return fmt.Errorf("failed to list access control rules: %w", err)
+	}
+
+	var mcpServerInstances v1.MCPServerInstanceList
+	if err := req.List(&mcpServerInstances, &client.ListOptions{
+		Namespace: system.DefaultNamespace,
+	}); err != nil {
+		return fmt.Errorf("failed to list MCP server instances: %w", err)
+	}
+
+	// Iterate through each MCPServerInstance and make sure it is still allowed to exist.
+	for _, instance := range mcpServerInstances.Items {
+		if instance.Spec.MCPCatalogName == "" {
+			// This instance points to a single-user server, so we don't need to worry about it.
+			continue
+		}
+
+		user, err := h.gatewayClient.UserByID(req.Ctx, instance.Spec.UserID)
+		if err != nil {
+			return fmt.Errorf("failed to get user %s: %w", instance.Spec.UserID, err)
+		}
```

Comment:
same question here

Reply:
the instances will get deleted when the servers get deleted because they have a DeleteRef to the server